dev/validate_bom__config.py

```diff
@@ -899,6 +913,179 @@ def add_files_to_upload(self, options, file_set):
+    if name is None:
+      return
+
+    script.append('hal -q --log=info config pubsub google enable')
+    script.append('hal -q --log=info config pubsub google subscription'
+                  ' add {name}'
+                  ' --project {project}'
+                  ' --json-path {credentials_path}'
+                  ' --subscription-name {subscription}'
+                  ' --template-path {template_path}'
+                  .format(name=name,
+                          project=project,
+                          credentials_path=credentials_path,
+                          subscription=subscription,
+                          template_path=template_path))
+
+  def add_files_to_upload(self, options, file_set):
+    """Implements interface."""
+    if options.pubsub_google_credentials_path:
+      file_set.add(options.pubsub_google_credentials_path)
+    if options.pubsub_google_template_path:
+      file_set.add(options.pubsub_google_template_path)
+
+
+class GcsPubsubNotficationConfigurator(Configurator):
+  """Controls external (to Spinnaker) GCS -> Google Pubsub config."""
+
+  def init_argument_parser(self, parser, defaults):
+    """Implements interface."""
+    add_parser_argument(
+        parser, 'gcs_pubsub_bucket', defaults, None,
+        help='The name of the bucket to create.')
+    add_parser_argument(
+        parser, 'gcs_pubsub_credentials_path', defaults, None,
+        help='The path to the credentials used to manipulate GCS and pub/sub.')
+    add_parser_argument(
+        parser, 'gcs_pubsub_project', defaults, None,
+        help='The name of the project that houses the bucket, topic, and subscription.')
+    add_parser_argument(
+        parser, 'gcs_pubsub_topic', defaults, None,
+        help='The name of the topic to create.')
+    add_parser_argument(
+        parser, 'gcs_pubsub_subscription', defaults, None,
+        help='The name of the subscription to create.')
+
+  def validate_options(self, options):
+    """Implements interface."""
+    if (options.gcs_pubsub_bucket is None
+        or options.gcs_pubsub_topic is None
+        or options.gcs_pubsub_project is None
+        or options.gcs_pubsub_credentials_path is None
+        or options.gcs_pubsub_subscription is None):
+      raise ValueError('Incomplete GCS -> Google pub/sub specification: '
+                       ' --gcs_pubsub_bucket="{}"'
+                       ' --gcs_pubsub_topic="{}"'
+                       ' --gcs_pubsub_project="{}"'
+                       ' --gcs_pubsub_credentials_path="{}"'
+                       ' --gcs_pubsub_subscription="{}"'
+                       .format(options.gcs_pubsub_bucket,
+                               options.gcs_pubsub_topic,
+                               options.gcs_pubsub_project,
+                               options.gcs_pubsub_credentials_path,
+                               options.gcs_pubsub_subscription))
+    if options.gcs_pubsub_subscription != options.pubsub_google_subscription_name:
```

Comment:
instead of raise ValueError call
raise_and_log_error(ConfigError(...))

The raise_and_log_error (from buildtool) is instrumented. The ConfigError will better categorize the source of problem to be configuration related ( a form of user error) as opposed to something else (a testing bug uncovered)

If I missed other errors thrown, this raise_and_log_error woudl apply there too. See the errors in buildtool/errors.py as the preferred choices. You can still use ValueError but categorizing them for trend analysis is more useful than precision since the errors are not actually handled.

Reply:
Done.